rust/src/lib.rs

```diff
@@ -3277,6 +3180,326 @@ pub unsafe extern "C" fn zcashlc_create_proposed_transactions(
+                    }
+                    TransactionDataRequest::SpendsFromAddress {
+                        address,
+                        block_range_start,
+                        block_range_end,
+                    } => FfiTransactionDataRequest::SpendsFromAddress {
+                        address: CString::new(address.encode(&network)).unwrap().into_raw(),
+                        block_range_start: block_range_start.into(),
+                        block_range_end: block_range_end.map_or(-1, |h| u32::from(h).into()),
+                    },
+                })
+                .collect(),
+        ))
+    });
+    unwrap_exc_or_null(res)
+}
+
+//
+// Tor support
+//
+
+/// Creates a Tor runtime.
+///
+/// # Safety
+///
+/// - `tor_dir` must be non-null and valid for reads for `tor_dir_len` bytes, and it must
+///   have an alignment of `1`. Its contents must be a string representing a valid system
+///   path in the operating system's preferred representation.
+/// - The memory referenced by `tor_dir` must not be mutated for the duration of the
+///   function call.
+/// - The total size `tor_dir_len` must be no larger than `isize::MAX`. See the safety
+///   documentation of pointer::offset.
+/// - Call [`zcashlc_free_tor_runtime`] to free the memory associated with the returned
+///   pointer when done using it.
+#[no_mangle]
+pub unsafe extern "C" fn zcashlc_create_tor_runtime(
+    tor_dir: *const u8,
+    tor_dir_len: usize,
+) -> *mut TorRuntime {
+    let res = catch_panic(|| {
+        let tor_dir = Path::new(OsStr::from_bytes(unsafe {
+            slice::from_raw_parts(tor_dir, tor_dir_len)
+        }));
+
+        let tor = crate::tor::TorRuntime::create(tor_dir)?;
+
+        Ok(Box::into_raw(Box::new(tor)))
+    });
+    unwrap_exc_or_null(res)
+}
+
+/// Frees a Tor runtime.
+///
+/// # Safety
+///
+/// - If `ptr` is non-null, it must point to a struct having the layout of [`TorRuntime`].
+#[no_mangle]
+pub unsafe extern "C" fn zcashlc_free_tor_runtime(ptr: *mut TorRuntime) {
+    if !ptr.is_null() {
+        let s: Box<TorRuntime> = unsafe { Box::from_raw(ptr) };
+        drop(s);
+    }
+}
+
+/// A decimal suitable for converting into an `NSDecimalNumber`.
+#[repr(C)]
+pub struct Decimal {
+    mantissa: u64,
+    exponent: i16,
+    is_sign_negative: bool,
+}
+
+impl Decimal {
+    fn from_rust(d: rust_decimal::Decimal) -> Option<Self> {
+        d.mantissa().abs().try_into().ok().map(|mantissa| Self {
+            mantissa,
+            exponent: -(d.scale() as i16),
+            is_sign_negative: d.is_sign_negative(),
+        })
+    }
+}
+
+/// Fetches the current ZEC-USD exchange rate over Tor.
+///
+/// The result is a `u128` containing a packed decimal:
+/// - Bits 16-23: Contains "e", a value between 0-28 that indicates the scale.
+/// - Bit 31: the sign of the Decimal value, 0 meaning positive and 1 meaning negative.
+/// - Bits 32-127: Contains the representation of the Decimal value as a 96-bit integer.
```

Comment:
```suggestion
/// - Bits 32-127: Contains the mantissa of the Decimal value as a 96-bit integer.
```
(I think the other instances are generated from this.)

Reply:
The language here was careful to use "representation" and not "mantissa", because mantissa means two different things depending on whether you are on the Rust or Swift side. However, this entire section is out-of-date, because I replaced the `u128` across the FFI with a `Decimal` C struct.